amethyst_ui/src/text_editing.rs

```diff
@@ -0,0 +1,360 @@
+use std::ops::Range;
+
+use clipboard::{ClipboardContext, ClipboardProvider};
+use unicode_normalization::{char::is_combining_mark, UnicodeNormalization};
+use unicode_segmentation::UnicodeSegmentation;
+use winit::{ElementState, Event, KeyboardInput, ModifiersState, VirtualKeyCode, WindowEvent};
+
+use amethyst_core::{
+    shrev::{EventChannel, ReaderId},
+    specs::prelude::{Join, Read, ReadStorage, Resources, System, WriteStorage},
+};
+use {Selected, TextEditing, UiText};
+
+/// System managing the keyboard inputs for the editable text fields.
+/// ## Features
+/// * Adds and removes text.
+/// * Moves selection cursor.
+/// * Grows and shrinks selected text zone.
+pub struct TextEditingInputSystem {
+    /// A reader for winit events.
+    reader: Option<ReaderId<Event>>,
+}
+
+impl TextEditingInputSystem {
+    /// Creates a new instance of this system
+    pub fn new() -> Self {
+        Self { reader: None }
+    }
+}
+
+impl<'a> System<'a> for TextEditingInputSystem {
+    type SystemData = (
+        WriteStorage<'a, UiText>,
+        WriteStorage<'a, TextEditing>,
+        ReadStorage<'a, Selected>,
+        Read<'a, EventChannel<Event>>,
+    );
+
+    fn run(&mut self, (mut texts, mut editables, selecteds, events): Self::SystemData) {
+        for text in (&mut texts).join() {
+            if (*text.text).chars().any(is_combining_mark) {
+                let normalized = text.text.nfd().collect::<String>();
+                text.text = normalized;
+            }
+        }
+
+        for event in events.read(
+            self.reader
+                .as_mut()
+                .expect("`UiKeyboardSystem::setup` was not called before `UiKeyboardSystem::run`"),
+        ) {
+            // Process events for the focused text element
+            if let Some((ref mut focused_text, ref mut focused_edit, _)) =
+                (&mut texts, &mut editables, &selecteds).join().next()
+            {
+                match *event {
+                    Event::WindowEvent {
+                        event: WindowEvent::ReceivedCharacter(input),
+                        ..
+                    } => {
+                        // Ignore obsolete control characters, and tab characters we can't render
+                        // properly anyways.  Also ignore newline characters since we don't
+                        // support multi-line text at the moment.
+                        if input < '\u{20}' {
```

Comment:
Maybe put all of these conditions into a `should_skip_char` or something, will probably make this easier to read